kpatch-build/create-diff-object.c

```diff
@@ -2408,10 +2436,45 @@ static void kpatch_create_intermediate_sections(struct kpatch_elf *kelf,
 		if (!strcmp(sec->name, ".rela.kpatch.funcs") ||
 		    !strcmp(sec->name, ".rela.kpatch.dynrelas"))
 			continue;
+#ifdef __powerpc__
+		/*
+		 * .rela.toc will be checked for dynrelas after processing
+		 * all sections. toc_rela(rela) might try to find an entry
+		 * which has already been converted into dynrela, if not
+		 *  checked at the end of section list.
+		 */
+		if (!strcmp(sec->name, ".rela.toc"))
+			continue;
+toc_sec_dynrela:
+		list_for_each_entry_safe(rela, safe, &sec->relas, list) {
+			if (!strcmp(sec->name, ".rela.toc")) {
+				for (i = 0, found = 0; i < off_idx; i++) {
+					if (toc_dynrela_off[i] != rela->offset)
+						continue;
+					found = 1;
+					break;
+				}
+				if (!found && rela->sym->sec)
+					continue;
+			} else {
+				if (!may_need_dynrela(rela))
+					continue;
+				if (!strcmp(rela->sym->name, ".toc")) {
+					/*
+					 * It's ok to have duplicate .tocoffset,
+					 * instead of linear search to eliminate
+					 * duplicates.
+					 */
+					if (rela->type == R_PPC64_TOC16_HA)
+						toc_dynrela_off[off_idx++] = rela->addend;
+					continue;
+				}
+			}
+#else
```

Comment:
The control flow is way too complex and hard to follow.

What if we add a `rela->needs_dynrela` bool, which gets set in the first loop which increments the `nr` variable?  Then we could move the `may_need_dynrela()` call there, and that would hopefully vastly simplify this code (e.g., no need for the `toc_dynrela_off` array or the spaghetti code flow).

Reply:
clever :). Will re-do the patch based on your suggestion.